### PR TITLE
Add debug trace output and CI build

### PR DESCRIPTION
**Pull request by Jordanwangxy**
1. Revised dgb_dump() function and callers
2. Revised github yml file to build 2 iar projects in one action file.
3. Added the private function not_enter_bootloader_mode_reset() to check other reset interrupts

### CODE REVIEW

**Review by JCordara-CoolIT — APPROVED** [review]

File: User/Inc/main.h [hunk 1]
```diff
@@ -47,6 +47,9 @@ extern "C" {
 
 /* Exported constants --------------------------------------------------------*/
 /* USER CODE BEGIN EC */
+extern const uint8_t dbg_info_j2app[];
```
Comment (PeteHeckel) [inline User/Inc/main.h:50]: Why are we defining these as global variables? Why can't they just be strings local to where they're called?

File: User/Src/main.c [hunk 2]
```diff
@@ -523,3 +569,8 @@ void assert_failed(uint8_t *file, uint32_t line)
   /* USER CODE END 6 */
 }
 #endif /* USE_FULL_ASSERT */
+
+void dbg_dump (uint8_t *pData, uint16_t Size)
+{
+  HAL_UART_Transmit(&huart2, pData, Size, 20);
```
Comment (PeteHeckel) [inline User/Src/main.c:575]: Could you specify the magic number with a macro indicating what it is / units?

File: User/Src/main.c [hunk 3]
```diff
@@ -177,6 +200,29 @@ int main(void)
   } // while(true)
 }
 
+/**
+  * @brief System Clock Configuration
+  * @retval None
+  */
+static bool not_enter_bootloader_mode_reset(bool adapter_app)
```
Comment (PeteHeckel) [inline User/Src/main.c:207]: Could you update this comment to be representative of the function? Also maybe re-think the name of the function; it's functionality isn't super clear to me based off the name.

File: User/Src/main.c [hunk 4]
```diff
@@ -177,6 +200,29 @@ int main(void)
...
+{
+  if ((__HAL_RCC_GET_FLAG(RCC_FLAG_LPWRRST)) ||
```
Comment (PeteHeckel) [inline User/Src/main.c:209]: Minor nit but this could just be a return statement instead of a conditional

**Review by PeteHeckel — CHANGES_REQUESTED** [review]
Minor notes, looks good mainly

File: User/Src/main.c [hunk 5]
```diff
@@ -572,7 +566,7 @@ void assert_failed(uint8_t *file, uint32_t line)
 }
 #endif /* USE_FULL_ASSERT */
 
-void dbg_dump (uint8_t *pData, uint16_t Size)
+void dbg_dump (uint8_t *pData)
 {
-  HAL_UART_Transmit(&huart2, pData, Size, 20);
+  HAL_UART_Transmit(&huart2, pData, sizeof(pData), 20);
```
Comment (JCordara-CoolIT) [inline User/Src/main.c:571]: Does sizeof work here? I would've expected strlen to be used.

I agree with Peter here as well that the 20 should be replaced by a macro. (e.g. DBG_UART_TIMEOUT_MS)

File: User/Src/main.c [hunk 6]
```diff
@@ -203,24 +199,23 @@ int main(void)
...
 #else
-      (__HAL_RCC_GET_FLAG(RCC_FLAG_BORRST)) ||
+  if (__HAL_RCC_GET_FLAG(RCC_FLAG_BORRST)) return false;
 #endif        
-      (adapter_app && (!__HAL_RCC_GET_FLAG(RCC_FLAG_SFTRST))))
-  {
-    return false;
-  }
   
   return true;
```
Comment (JCordara-CoolIT) [inline User/Src/main.c:220]: I might suggest re-writing this function as the following:
```c
static bool is_app_sw_reset(void)
{
  return !(
    (__HAL_RCC_GET_FLAG(RCC_FLAG_LPWRRST)) ||
    (__HAL_RCC_GET_FLAG(RCC_FLAG_WWDGRST)) ||
    (__HAL_RCC_GET_FLAG(RCC_FLAG_IWDGRST)) ||
    (__HAL_RCC_GET_FLAG(RCC_FLAG_PORRST)) ||
#ifdef RS485_ADAPTER
    (__HAL_RCC_GET_FLAG(RCC_FLAG_OBLRST)) ||
#else
    (__HAL_RCC_GET_FLAG(RCC_FLAG_BORRST))
#endif
  );
}
```

**Review by JCordara-CoolIT — CHANGES_REQUESTED** [review]
Just a few more requests to finalize the changes

**Review by PeteHeckel — APPROVED** [review]

**Review by ChristieNgan — APPROVED** [review]

**Pull request merged**